### ISSUE TITLE
Spike: improve bazel experience in the repo

### ISSUE THREAD

**Issue by viovanov** [cloudfoundry-incubator/kubecf#545]
**Is your feature request related to a problem? Please describe.**
Getting started with bazel is hard.

**Describe the solution you'd like**
- perhaps a makefile to wrap useful bazel targets
- comments in the bazel build files and rules
- FAQ section

**Comment by mudler** [comment]
Very hard to achieve imho. I've already expressed my concerns, but I will add also my thoughts here on the subject.

- I see this issue a bit countersense, coming up with ways to improve its "daily" experience it just highlights that it's not good, and it is a failure (my personal point of view)  my experience with bazel

**Comment by mudler** [comment]
Very hard to achieve imho, I have already expressed my thoughts on the subject already, but I'll add them here too so everyone is free to debate with me in the public :smile:  :

- I see this issue a bit a countersense. Having to come up with ways to "simplify it" shows that bazel failed in this area. No matter how you will hide the complexity, we will have to face it on "daily" basis. And given that our team is not acquainted with it, it becomes even cumbersome. That is my personal opinion with the experience I had with it.
- Creating a Makefile to wrap it, sounds pretty much that we don't need bazel. Having two ways to building things on the same project is a red-flag for me, because we could just have one that works, and is really understood by the involved parties. Bazel looks like a top-down decision that (as far as I know) I'm not the only one that disagrees with.  Usually one have to support their decision to *adopt* new things, not the opposite way around.
- There are comments, even README in the build targets, that doesn't help much. Extending the docs about them, again, shows to me that it failed very hard in usability. I expect also to have docs that are human readable from bazel itself (yes, might sound hard, but I don't find them helpful)
- To be honest: I see working on this again as a wasted engineering cycles. "Wrapping up" will make things even more complex, and even harder to understand.
- I would expect a "novel" build system to give me more than just restrictions, at least should give me something back for the effort I put on it. There is no real catch of using bazel for me (again, personal opinion, but if we don't have opinion on softwares, means we are not using it :smile:  ) except the complexity built around it, and docs that are very difficult to follow. There are no even consistent examples that you can base your work on.
- I dislike Python (again, personal opinion) and I see the bazel DSL very strict (that doesn't mean necessarly a bad thing) but at the same time it's just readable to those wrote that DSL. There was no Human-Computer-Interaction goal in mind. I think the computer science literature on the subject is well structured, and gives a lot of inputs. I could cite more articles, but I would refrain me to do so, I will just cite the [most famous one](https://www.nngroup.com/articles/usability-101-introduction-to-usability/).
- I expect in the long run to be more proficient with it, but that doesn't mean that it is a success story. Instead it looks to me we are just exploring only one way to do things - and we keep going with it neverthless

**Comment by manno** [comment]
From the discussion before, I remember there was also some interest in re-organizing the ops files by function instead of by instance_group.
Is this part of this issue?

**Comment by f0rmiga** [comment]
@manno No. There should be a different story for that (not sure if it was created already).

**Comment by jimmykarily** [comment]
I'll not repeat what Ettore said, I'll just say I agree. I think bazel if (probably?) solving some problem but not a problem we have or a problem we should ever have in kubecf.

Building kubecf artifacts is and should remain a fast and simple process. It shouldn't need distributing the build workload in multiple machines, it shouldn't need sophisticated caching, it shouldn't need to check a huge list of dependencies before the build process starts because the build process should just take 3 seconds.

I'm not fluent in bazel syntax and idiosyncrasies but reading about bazel benefits (e.g. [here](https://blog.mozilla.org/nfroyd/2019/10/28/evaluating-bazel-for-building-firefox-part-1/)) I don't see any feature that makes it useful for kubecf.

Summing up, I have to agree, the best way to improve the bazel experience in my opinion is to remove bazel. If you think replacing bazel isn't worth the time and effort, you need to see how much time we loose every time we need to do a simple thing in bazel. We would have replaced it 3 times by now.

**Comment by jandubois** [comment]
Just for the record: I'm only doing some preliminary exploration to further the discussion, as we have not reached agreement yet, but @viovanov wanted me to move the issue to the project for this.

**Comment by jandubois** [comment]
# Replacing bazel with shell scripts

## Prior feedback from internal SUSE mailing list

With the exception of Thulio and Vlad all SUSE team-members that voiced an opinion were really negative about using bazel. From the internal email thread:

Dimitris:
>I spent an unreasonable amount of time trying to do simple things in bazel and most of the time I ended up finding workarounds just to avoid the deep dive in bazel docs.

Ettore:
>Our use case of bazel seems to be very small related to the features that it provides. At the end, you cannot even "steal" examples from real-use cases, as they are completely off from our usage.

Victor:
>It does it as a big black box, instead of a toolbox of little executables with its own manpages and docs.

Mario:
>If you don't need the docker image caching features, I guess it's just another overly complicated build system.

Mark:
>I too find the bazel documentation terrible: it's extremely non-obvious where the correct docs are. [...] We're also using it in a way it's not suited for: it's good for building things, but not so much running things [...].  Every time we're doing that it just feels like we're fighting against what it wants.

Me:
>I also dislike the way the whole bazel build system is opaque. You cannot run with a --dry-run option to see what it will execute; you can't even see what it does when it builds things because many actions are executed internally by bazel itself.

## Examples where bazel has gotten in the way

We can already see that bazel slows us down significantly. Look at PR #149 "Annotate operator chart url in Chart.yaml": almost 40 lines of changes over 4 different files, just to write the operator URL to a file in the helm chart.

Another example is PR #557 "Print image list script for kubecf manifest", later superseded by #571. It had been worked on and reviewed by multiple people, but to make it work properly with bazel it had to be redone by Thulio, who is our only bazel expert.

As Mark mentioned above, we mostly use bazel not for building, but for running. But it isn't good enough for that either, so we have PR #573 "Add a Makefile with easy to discover targets" to wrap bazel targets with a Makefile. Wrapping a dependency tracking build system with yet another dependency tracking build system, that we again just use as a launcher.

## Kubernetes is getting rid of bazel

Even kubernetes itself is getting rid of bazel: [Please remove bazel based build infrastructure](https://github.com/kubernetes/kubernetes/issues/88553):

>With the switch to go modules/cache, most of the reasons why bazel was used (speed up compilations) are gone.

The reason they were using bazel doesn't even apply to us: we don't compile code for kubecf, so we don't need caching.

But the problems they run into sound extremely familiar:

>The #bazel channel is filled with questions like, "How do I do this thing that should be moderately simple to do if it wasn't wrapped in bazel?"

>This is a huge maintainability burden, with our limited set of bazel experts being some of the most bandwidth-constrained contributors in the project.

And the release engineering team seems fully on board with getting rid of bazel:

>@kubernetes/release-engineering is on board to help tear this down. Just let us know which threads we need to start pulling on to make this happen.

## Alternative implementation in bash

I've created a proof of concept, implementing a couple of the bazel targets in simple shell scripts. See PR #583 "Build and run kubecf without bazel" for the code. It is neither robust nor complete; it only exists to illustrate the point:

```
$ time make kubecf-build
./scripts/kubecf-build.sh
Successfully packaged chart and saved it to: output/kubecf-v0.0.0-ab6026b.tgz

real	0m2.420s
```

It builds the complete product from scratch in 2-3 seconds. There is no caching, no dependency tracking; everything is built from scratch every single time. And it is just 50 lines of simple shell script that I expect all team members to understand easily.

The files generated are identical to the ones created by `bazel build //deploy/helm/kubecf`, except for the sort order of the keys in 2 yaml files because I write them using `ruby` instead of `yq`.

I've prototyped a couple of other scripts as well to show the full kubecf build/test cycle can be done with very little code:

```
$ VM_DRIVER=virtualbox MINIKUBE_EXTRA_OPTIONS=--registry-mirror=http://dockerhub.home:5000 make minikube-start
$ make cf-operator-apply
$ pod-status -w
$ make kubecf-build
$ make kubecf-apply
$ pod-status -w
$ make testing-smoke
$ kubectl logs --follow --namespace kubecf \
>         --selector "quarks.cloudfoundry.org/qjob-name in (kubecf-smoke-tests)" \
>         --container smoke-tests-smoke-tests
Running smoke tests...
Running binaries smoke/isolation_segments/isolation_segments.test
smoke/logging/logging.test
smoke/runtime/runtime.test
[1585700822] CF-Isolation-Segment-Smoke-Tests - 4 specs - 4 nodes SSSS SUCCESS! 17.960221513s
...
```

Looking at this experiment, I now firmly believe that we should replace our bazel rules with a couple of shell scripts.

If we implement some additional mechanism to allow downloading and using specific binary versions of tools used by the scripts (helm, jq) then we should have all the functionality provided by the current bazel setup, but in a format that can be maintained by all team members and is easy to understand by external contributors.

Please let me know if we are using any other bazel feature, that we would be losing by making this switch?

**Comment by viovanov** [comment]
We have many (types) of dependencies
- docker images
- bosh releases
- helm charts
- ops files
- scripts
- tests

And a lot of tooling
- helm
- kubectl
- kind
- k
- yq
- jq

This project is not meant to have constant code churn.
It's meant to provide a rock-solid integration of Cloud Foundry components for Kubernetes.
That is the goal. 
Bazel helps us pin down dependencies and provides us with reproducible builds using well defined rules.
Again, this is the absolute goal of this project.

I am not convinced that implementing bazel rules is as difficult as the team says it is. There is a learning curve, like any other tool, but steeper.
I also don't think it's a good idea to give it up, given that it does everything we need it to.
No, I really don't want to give up on Bazel, I think it's just beginning to show its potential for integrating complex projects.

Implementing everything that bazel provides in shell is a bad idea.
We use shell scripts for small units of work, of course, but shell scripts are not good for building complex frameworks and dependency management.

We'll cut corners and implement something that we don't need. The build system for kubecf is there, and working well.

I am open to the idea of removing dev environments from kubecf's responsibilities though.

#149 - yes, we are building helm rules for bazel (I actually find this work awesome, and it should live in a repo of its own)
#557 - nothing was "redone" - @f0rmiga picked up a spike from @rohitsakala (it was never meant to be the final implementation, just proof it could be done).
#573 - I guess it's my fault for trying to help; was simply trying to empathize, and essentially provide a list of aliases, nothing more; I think calling it a "wrapper" is inaccurate.

About Kubernetes - I am not sure that's relevant to us - afaict they haven't made a decision yet, and they don't have such a varied dependency list.
There are many success stories out there for Bazel:
- [SpaceX](https://www.youtube.com/watch?v=t_3bckhV_YI)
- [TensorFlow](https://www.youtube.com/watch?v=Rw-KrbfyABQ) 
- [Dropbox](https://www.infoq.com/news/2020/01/dropbox-bazel-ci/)

**Comment by qu1queee** [comment]
My 2 cents on this:

- Hard to give an opinion on the topic while I think the hassle comes when developing bazel rules or similar. I haven´t done that yet, and my interactions with the build tool is restricted to a couple of `targets`, mainly `bazel build //deploy/helm/kubecf`.
- The learning curve is definitely high, I can immediately see this when reading parts of the bazel config, but this again, does not have an impact on users that only run the main `targets`. Unless is debugging a failed target, I have get into this situation, which can be tedious, but probably as with any debugging.
- For the dependency management, I think Bazel does a good job, e.g. the [cf-deployment](https://github.com/cloudfoundry-incubator/kubecf/blob/master/def.bzl#L144-L161) base YAMLs. It would be a nice exercise to do this in shell(duable but probably not so clean).
- It seems the amount of bazel targets is high. As an end-user I´m probably mainly interested on packaging the charts and consuming them, which boils down to one or two targets(e.g._bazel build //deploy/bundle:cf_operator_chart_ and _bazel build //deploy/helm/kubecf:version_). It might be a good idea to revisit the amount of targets that this repo supports, e.g.  `bazel query //dev/...` vs `bazel query //deploy/...`, and reconsider if all of them are needed, or if some of them could be implemented differently.

**Comment by f0rmiga** [comment]
I'll push back on some of the mentions:

> I spent an unreasonable amount of time trying to do simple things in bazel and most of the time I ended up finding workarounds just to avoid the deep dive in bazel docs.

When did we start thinking that deep-diving into docs is a bad thing that should be avoided? Bazel docs may not be the best doc ever written, but it's there!

> Our use case of bazel seems to be very small related to the features that it provides. At the end, you cannot even "steal" examples from real-use cases, as they are completely off from our usage.

We are not after copy-pasting things, as far as I can tell. The objective is to understand the value of deterministic builds and to bring that philosophy to KubeCF. The "black-box," as some like to call the rules, is doing it on a sandboxed filesystem and without the interference of the terminal session environment, i.e. the user environment variables don't cause side-effects in the build.

> It does it as a big black box, instead of a toolbox of little executables with its own manpages and docs.

I agree that the rules should have better documentation, but they are very well broken into small rules that, in most of the cases, do one and one job. Whenever we break the rules, we should rethink and fix, not get rid of the whole build system because some people don't understand it.

> If you don't need the docker image caching features, I guess it's just another overly complicated build system.

We *can* do image caching, and I'd love it (I guess everyone would). Like Vlad mentioned, "No, I really don't want to give up on Bazel, I think it's just beginning to show its potential for integrating complex projects."
We can create tasks for implementing it in a similar way that Gazelle does for Go rules.

> I too find the bazel documentation terrible: it's extremely non-obvious where the correct docs are. [...] We're also using it in a way it's not suited for: it's good for building things, but not so much running things [...]. Every time we're doing that it just feels like we're fighting against what it wants.

Bazel is well-suited for *running* things. It's more work because we want to eliminate as much as possible the dependencies on the user environment to run, e.g. tracking the kind, minikube, kubectl and helm binaries.

As soon as we get rid of this strict dependency tracking, we will start having all sorts of issues due to version incompatibilities that we cannot completely foresee. As long as we keep track of the versions in the way we do, we can point people to the versions we are using and testing *many* times a day!

> I also dislike the way the whole bazel build system is opaque. You cannot run with a --dry-run option to see what it will execute; you can't even see what it does when it builds things because many actions are executed internally by bazel itself.

I also dislike that, but we should be working together to overcome this, perhaps with creative solutions that we could share with the community!

---

Lastly, as mentioned by Enrique, we don't expect everyone to touch the Bazel rules all the time, and KubeCF development shouldn't be mostly focused on ops-files.

**Comment by mudler** [comment]
The toolchain doesn't have to define a project **goal** or **scope**. The toolchain have to fit the project, not the other way around. The project should be completely disjointed by it. It seems instead we are bending KubeCF to the toolchain being adopted right now.
 
Let's take a step back and look at what bazel is good at: big codebase that needs to be compiled and needs caching. Now lets take a look through KubeCF: a set of YAML files.

Taking the right tool for the right job is more important in my honest opinion. Clearly bazel doesn't fit, as it does much more than what we need and @jandubois PR shows that actually we could live without it (PR needs improvement, yes, like pinning dependencies, but we don't have so many: `k` or `kind` for example, are not necessary to build a helm chart).

The use cases you mentioned are really good for big C, or C++ projects, where compilation times are drastically high ( I have some experience with it ) and bazel shared compilation mechanism can help a lot: but not at YAML packaging. 

Tensorflow it's an amalgams of C++/C code with Python on top. And the examples given are totally out of our context (feel free to push me back, but I still believe that KubeCF its not something that have to be compiled, and you cannot leverage the "distributed compilation" or "caching" features that bazel advertize). I would suggest you to go through the Kubernetes issue Jan mentioned, there you can see the full rationale.

For example, the dependencies can be be tracked as well in a single docker image. The use-cases you mention here (pinning dependencies, and so on) doesn't require a toolchain that is meant for different scopes. Even docker can be used to containerize and pin dependencies to run workloads, I don't see our use case requiring so much complex tooling that needs a build system that was clearly designed for other uses cases in mind.

Docs are really important, they shouldn't "just exist" but they have to be clear and well scoped. If you cannot take the example they provide and run it locally with success they have important flaws that makes the steep learning curve even higher without a good reason.

To end my dissertion here, and I won't add more as it looks a bit of waste of energy ( :smile: ) : It's not just "some" people not liking it as shown in https://github.com/cloudfoundry-incubator/kubecf/issues/545#issuecomment-606976345. Apparently a decision was already taken, and we have to live with it.

With a metaphor that is really known (maybe only in my country): we are killing a mosquito with a (broken)bazooka.

**Comment by jandubois** [comment]
It looks like the strict tracking of tool dependencies was the last remaining feature of the bazel rule system that was missing from my proof-of-concept in #583. I've implemented this too now, to prove that it is not as hard as some seem to think.

I strongly believe that the scripts now provide all the functionality we use from bazel, while improving on several of the shortcomings:

### Essential features provided by both bazel and the shell based build system

* Fast build time

* Tracking of all code dependencies

* Tracking of many tool dependencies

### Essential features only provided by the shell based build system

* Understandable and maintainable by all team members; easy to pick up by contributors

* Transparent execution that can be traced and debugged; steps can be executed manually and intermediate results can be inspected; parts are reusable outside the build framework

* Serves as executable documentation how to perform various tasks; less likely to be obsolete than pure docs

### Essential features only provided by the bazel rules based system

None that I'm ware of (and I've reviewed both the previous comments here, as well as the prior internal SUSE mailing list thread). If I missed anything substantial (that we actually use in this project), please let me know!

### Conclusion

The long and steep learning curve has been acknowledged by everyone. But the benefit that justifies the required big investment of time remains unclear.

The claim that most people don't need to understand how the project is being built feels ridiculous. We make changes to build and test scripts all the time, and everyone, including people working on CI and QA need to be able to understand how this works, and make changes easily and with confidence.

The essence of the shell scripts already exists inside the bazel rules system. The scripts are just obfuscated as templates and not reusable outside of bazel. It is not possible to manually execute them step by step for debugging, or at even get a trace of what happened.

Having accessible scripts also provides documentation. The question from @cweibel above is a perfect example. If we already had [scripts/test-smoking.sh](https://github.com/cloudfoundry-incubator/kubecf/blob/f226eb8/scripts/testing-smoke.sh) then he would have likely found the answer by himself and didn't have to ask.

How can we justify **not** switching to a system that would work so much better for us?

**Comment by dragonchaser** [comment]
Just to add my 2 cents here:
I think bazel is a bit over-engineered for this purpose. As I see it it overcomplicates things. I am in favour of @jandubois suggestion to reimplement it in bash.

**Comment by viovanov** [comment]
We've decided to remove Bazel as a build system for kubecf.
Closing this in favor of #749
